lua/nvim-tree/actions/tree/modifiers/expand.lua

```diff
@@ -27,20 +27,65 @@ local function expand(node)
   end
 end
 
+---@param should_descend fun(expansion_count: integer, node: Node): boolean
+---@return fun(expansion_count: integer, node: Node): boolean
+local function limit_folder_discovery(should_descend)
+  return function(expansion_count, node)
+    local should_halt = expansion_count >= M.MAX_FOLDER_DISCOVERY
+    if should_halt then
+      return false
+    end
+
+    return should_descend(expansion_count, node)
+  end
+end
+
 ---@param expansion_count integer
 ---@param node Node
 ---@return boolean
-local function should_expand(expansion_count, node)
+local function descend_until_empty(expansion_count, node) --luacheck: ignore 212
```

Comment:
I like the way you've applied the consistent `limit_folder_discovery` to both modes of operation.

However, ignoring checks is not acceptable, and this also flags an unused-local warning. The first argument is necessary to keep the same function signature however we can ignore it.

```suggestion
local function descend_until_empty(_, node)
```